### PR TITLE
Enhancement: Retain Original User Description and Title in PRs

### PR DESCRIPTION
**Pull request by zmeir**
## PR Type:
Enhancement

___
## PR Description:
This PR introduces a new feature that allows users to retain their original PR description and title. This can be useful when users want to add extra information that the PR Agent can't create with the describe command, such as feature requirements, user story, design spec, etc. The feature is controlled via two configuration toggles which are disabled by default, allowing users to opt-in if they want this new behavior. The changes are mainly implemented in the GitProvider classes and the PRDescription class.

___
## PR Main Files Walkthrough:
`pr_agent/git_providers/bitbucket_provider.py`: Refactored BitbucketProvider to inherit from GitProvider. Removed the get_pr_description method and replaced it with get_pr_description_full.
`pr_agent/git_providers/git_provider.py`: Added new methods get_pr_description_full, get_pr_description, and get_user_description to the GitProvider abstract base class.
`pr_agent/git_providers/github_provider.py`: Removed the get_pr_description method and replaced it with get_pr_description_full in GithubProvider.
`pr_agent/git_providers/gitlab_provider.py`: Removed the get_pr_description method and replaced it with get_pr_description_full in GitlabProvider.
`pr_agent/git_providers/local_git_provider.py`: Replaced the get_pr_description method with get_pr_description_full in LocalGitProvider.
`pr_agent/tools/pr_description.py`: Added user_description as an attribute of the PRDescription class. Modified the _prepare_pr_answer method to include the user description in the PR body if the keep_user_description setting is enabled. Also added the option to keep the original user title.
`pr_agent/settings/configuration.toml`: Added two new configuration options add_original_user_description and keep_original_user_title under the pr_description section.

### CODE REVIEW

**Comment by idavidov** [comment]
/ask what this pr going to fix

File: pr_agent/git_providers/bitbucket_provider.py [hunk 1]
```diff
@@ -6,12 +6,11 @@
...
 from ..config_loader import get_settings
-from .git_provider import FilePatchInfo
+from .git_provider import FilePatchInfo, GitProvider
 
 
-class BitbucketProvider:
+class BitbucketProvider(GitProvider):
     def __init__(self, pr_url: Optional[str] = None, incremental: Optional[bool] = False):
         s = requests.Session()
         s.headers['Authorization'] = f'Bearer {get_settings().get("BITBUCKET.BEARER_TOKEN", None)}'
```
[the remark this answers is not in the document]
Comment (zmeir) [reply pr_agent/git_providers/bitbucket_provider.py:16]: Not relevant to this PR

File: pr_agent/git_providers/git_provider.py [hunk 2]
```diff
@@ -82,9 +82,26 @@ def get_user_id(self):
...
+            return clip_tokens(description, max_tokens)
+        return description
+
+    def get_user_description(self):
+        description = (self.get_pr_description_full() or "").strip()
+        if not description.startswith("## PR Type"):
+            return description
+        if "## User Description:" not in description:
+            return ""
+        return description.split("## User Description:", 1)[1].strip()
```
[the remark this answers is not in the document]
Comment (zmeir) [reply pr_agent/git_providers/git_provider.py:103]: It is handled in line 98: `description = (self.get_pr_description_full() or "").strip()`

File: pr_agent/settings/configuration.toml [hunk 3]
```diff
@@ -24,6 +24,7 @@ extra_instructions = ""
 
 [pr_description] # /describe #
 publish_description_as_comment=false
+keep_user_description=false
 extra_instructions = ""
```
[the remark this answers is not in the document]
Comment (zmeir) [reply pr_agent/settings/configuration.toml:28]: Good idea, but I see no other configuration option has a comment, so I'm wondering if this is the right place for it? Perhaps the issue of documenting all configuration options should be a addressed in a separate PR?  
@idavidov let me know what you think.

**Comment by idavidov** [comment]
sorry for the noise, i am using this PR to get some pictures for article i am working on :)

**Comment by zmeir** [comment]
You can take a look at https://github.com/zmeir/pr-agent/pull/5 for examples of using this feature. Look at the edit history for a full rundown of the testing (feel free to use the PR for your testing as well):
1. Created a PR with no description and ran describe from the cli with the configuration disabled to see it has no ill-effect
2. Ran describe again on the PR, still with the configuration disabled.
3. Cleared the description for the next batch of testing
4. Ran describe with the configuration enabled to see it works the same (since there's no user description to append)
5. Ran describe again with the configuration enabled to see it works the same (there is no user description section in the PR description so it shouldn't append anything)
6. Cleared the description and added some custom user description of my own
7. Ran describe with the configuration enabled to see it keeps the user description and appends it to the generated one
8. Ran describe again with the configuration enabled to see it keep __just__ the user description and adds it to the generated one (instead of appending all of the previously generated description to the new one, resulting in duplications).

**Comment by mrT23** [comment]
@idavidov Notice that just for "practicing", you can set
```
[config]
publish_output=true
verbosity_level=2 # 0,1,2
```

this may not be useful for pictures (😄 ), but its good when developing\modifying features without noise

**Comment by ilyadav** [comment]
Thank you for tip,  I wasn't aware about this feature

File: pr_agent/git_providers/git_provider.py [hunk 4]
```diff
@@ -82,8 +82,25 @@ def get_user_id(self):
         pass
 
     @abstractmethod
-    def get_pr_description(self):
-        pass
+    def get_pr_description_full(self) -> str:
+        pass
+
+    def get_pr_description(self) -> str:
+        from pr_agent.config_loader import get_settings
```
Comment (mrT23) [inline pr_agent/git_providers/git_provider.py:89]: can you move the imports to the top ?

**Comment by mrT23** [comment]
Hi @zmeir , some questions:

1. just to make sure I understand: the original description would be additive, right ?
meaning we will present both the original description, and the AI-generated description

**Comment by zmeir** [comment]
Hi @mrT23,

You can see a detailed description of how I tested this PR in this comment: https://github.com/Codium-ai/pr-agent/pull/223#issuecomment-1685744238 and examples for its output in this PR: https://github.com/zmeir/pr-agent/pull/5

For instance, if this was my original description:  
image  
Then after running the describe command, I got this:  
image

To answer your questions:
1. Yes. More specifically I like to think of how this feature handles 4 distinct when you call the describe command:
   1. There is no description at all - The agent generates the description without change as it does today (like this feature is disabled)
   2. There is existing user description (not generated by the agent) - The agent generated its description and adds the user description in a `## User Description:` section
   3. There is existing agent description but without an added user description (because the feature was disabled or there simply was no user description to add) - The agent simply generates its description without adding the existing description (this is the same behavior as without this feature - which makes sense because you don't want to append the agent's previous description to its new description)
   4. There is existing agent description with added user description - This is the most interesting case. In this case the agent will generate its description, and append to it only the user description from the existing agent description. This ensures that you can keep running the describe command repeatedly, and the agent will still keep your original description, and it also allows you to edit it and keep the changed between invocations of the describe command.

   Good point about the configuration name - I'll change it.

2. Sure, it goes like this:
   1. The first condition `if not description.startswith("## PR Type")` checks if the existing description __wasn't__ generated by the pr-agent, and in that case it just return it as-is.
   2. The second condition `if "## User Description:" not in description checks if the existing description __was__ generated by the pr-agent, but it doesn't contain the user description, and in that case it returns nothing (empty string) because it means there is no user description (this is the case when you ran the describe command on a PR with no description at all, or with the configuration for this feature disabled)
   3. Lastly, if all the previous conditions aren't met, then it means the existing description that was generated by the pr-agent contains the original user description. It appears after the `## User Description:` section of the description text because that's how `PRDescription._prepare_pr_answer` generates the markdown text for the description, and we make sure it's the last section in the pr-agent generated description by adding it last to the `data` dictionary after all the generated description data was loaded with `data = load_yaml(self.prediction.strip())`
   
   Another way to write this same logic it to inverse the conditions, then maybe the logic will be clearer?
   ```python
    def get_user_description(self) -> str:
        description = (self.get_pr_description_full() or "").strip()
        if description.startswith("## PR Type"):
            if "## User Description:" in description:
                user_description = description.split("## User Description:", 1)[1].strip()
            else:
                user_description = ""
        else:
            user_description = description
        return user_description
   ```
    Let me know what you prefer. Personally I don't like the nesting. Anyway, I'll add comments to explain the logic.

3. That's your call :) If you think it's useful I'll add it. From what I saw the generated title is usually spot-on so I never needed to change it.

**Review by mrT23 — APPROVED** [review]

**Comment by mrT23** [comment]
@zmeir 
looks great

can i merge ?

**Comment by zmeir** [comment]
> @zmeir looks great
> 
> can i merge ?

Thanks 😃 

No objections from me 👍🏽

**Pull request merged**